pkg/shed/db.go

```diff
@@ -73,84 +85,307 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+			return ErrNotFound
+		}
+		value, err = i.Item().ValueCopy(value)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		db.metrics.FirstFailCount.Inc()
+	}
+	return key, value, err
+}
+
+// Last retuns the last key matching the given prefix.
+func (db *DB) Last(prefix []byte) (key []byte, value []byte, err error) {
+	db.metrics.LastCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = true
+		o.PrefetchSize = 1024
+		o.Reverse = true   // iterate backwards
+
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		// get the next prefix in line
+		// since leveldb iterator Seek seeks to the
+		// next key if the key that it seeks to is not found
+		// and by getting the previous key, the last one for the
+		// actual prefix is found
+		nextPrefix := incByteSlice(prefix)
+		l := len(prefix)
+
+		if l > 0 && nextPrefix != nil {
+			// If there is a no key which starts which nextPrefix, badger moves the
+			// cursor to the previous key (which should be our key).
+			i.Seek(nextPrefix)
+			if bytes.HasPrefix(i.Item().Key(),prefix) {
+				key = i.Item().Key()
+				value, err = i.Item().ValueCopy(nil)
+				if err != nil {
+					return err
+				}
+			} else {
+				// If there is a key which starts with nextPrefix, we do reverse Next() to
+				// reach our key and pick that up.
+				i.Next()
+				if bytes.HasPrefix(i.Item().Key(),prefix) {
+					key = i.Item().Key()
+					value, err = i.Item().ValueCopy(nil)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		if key == nil {
+			return ErrNotFound
+		}
+		return nil
+	})
+	if err != nil {
+		db.metrics.LastFailCount.Inc()
+	}
+	return key, value, err
 }
 
-// NewIterator wraps LevelDB NewIterator method to increment metrics counter.
-func (db *DB) NewIterator() iterator.Iterator {
-	db.metrics.IteratorCounter.Inc()
-	return db.ldb.NewIterator(nil, nil)
+// GetBatch get a new badger transaction to be used for multiple atomic operations.
+func (db *DB) GetBatch(update bool) (txn *badger.Txn) {
+	db.metrics.GetBatchCount.Inc()
+	// set update to true indicating that data will be added/changed in this transaction.
+	return db.bdb.NewTransaction(update)
 }
 
-// WriteBatch wraps LevelDB Write method to increment metrics counter.
-func (db *DB) WriteBatch(batch *leveldb.Batch) (err error) {
-	err = db.ldb.Write(batch, nil)
+// WriteBatch commits the badger transaction after all the operations are over.
+func (db *DB) WriteBatch(txn *badger.Txn) (err error) {
+	db.metrics.WriteBatchCount.Inc()
+	err = txn.Commit()
 	if err != nil {
-		db.logger.Debugf("could not writing batch. Error: %s ", err.Error())
-		db.metrics.WriteBatchFailCounter.Inc()
+		db.metrics.WriteBatchFailCount.Inc()
 		return err
 	}
-	db.metrics.WriteBatchCounter.Inc()
 	return nil
 }
 
-// Close closes LevelDB database.
+// Close shuts down the badger DB.
 func (db *DB) Close() (err error) {
-	close(db.quit)
```

Comment:
i wonder, we dont need this anymore?

Reply:
Where is this?   we have a Close() at the end of the file which doesn't look like this.

```
// Close shuts down the badger DB.
func (db *DB) Close() (err error) {
	db.logger.Tracef("database closed with path %s", db.path)
	return db.bdb.Close()
}
```